### PR TITLE
Import changes from upstream prelude

### PR DESCRIPTION
**Pull request by avdv**
This PR introduces a few changes from upstream, mostly in order to reduce the amount of changes between the
current prelude and ourselves to make it easier to upstream, and to benefit from upstream changes.

- **Dedupe attribute lists of deps**
- **Use `at_argfile` helper**
- **Remove left-over comment**
- **Add performance optimisations for handling package DBs**
- **Move package db selection before the loop**
- **Use attr_deps_haskell_toolchain_libraries**
- **Remove `exposed_package_libs`**
- **Reformat**
- **Convert cxx toolchain LinkerType to enum**
- **Fix haskell_ghci() for multi-version GHC**
- **Rename relative args -> args**

### CODE REVIEW

File: haskell/compile.bzl [hunk 1]
```diff
@@ -356,7 +354,26 @@ def get_packages_info2(
...
+        elif use_empty_lib:
+            db = lib.empty_db
+        else:
+            db = lib.db
+        packagedb_set[db] = None
+        hidden_args = cmd_args(hidden = [
+            lib.import_dirs.values(),
+            lib.stub_dirs,
+            # libs of dependencies might be needed at compile time if
+            # we're using Template Haskell:
```
Comment (aherrmann) [inline haskell/compile.bzl]: Just to be sure, this is not used in our fine grained module compile actions, correct? We don't want to depend on libraries in that case.

File: haskell/compile.bzl [hunk 2]
```diff
@@ -356,7 +354,26 @@ def get_packages_info2(
...
+            lib.libs,
+        ])
+        exposed_package_args.add(hidden_args)
+        packagedb_args.add(hidden_args)
```
Comment (aherrmann) [inline haskell/compile.bzl]: IIUC this is a change compared to before. Before the libraries were only added to `exposed_package_args`. Now they're also added to `packagedb_args`. Why is that necessary?

**Review by aherrmann — APPROVED** [review]
Looks good, but some questions about the libraries handling. I'd like to make sure we're not inadvertently introducing dependencies on link actions to metadata or compile actions.

File: haskell/compile.bzl (hunk 1 above)
Comment (avdv) [reply haskell/compile.bzl]: Yes, good catch. I was thinking about adding a flag to enable / disable this behavior to add backward compatibility for vanilla ghc but I want to tackle that separately. Removed.

\edit: Just removed the comment. The libs are no longer needed for TH, but are needed for the `exposed_package_args`.

File: haskell/compile.bzl (hunk 2 above)
Comment (avdv) [reply haskell/compile.bzl]: Right, this is not needed. Removed.
